### PR TITLE
Feature/baseHRC1155

### PR DESCRIPTION
**Pull request by AnnRiera**
Base HRC1155 structure and changes in HRC721 related to BaseToken abstract class.

### CODE REVIEW

File: src/base-implementation.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
+import { Wallet } from '@harmony-js/account'
```
Comment (bzuker) [inline src/base-implementation.ts:1]: I would call this file `base-token.ts` and remove the abstract base token created

File: src/base-implementation.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    this.name = BaseError.name
+    this.type = type
+    this.code = code
+    this.data = data
+
+    Error.captureStackTrace(this, this.constructor)
+  }
+}
+
+export class BaseToken extends Base {
```
Comment (bzuker) [inline src/base-implementation.ts:24]: let's make this class abstract and not extend anything

File: src/base-implementation.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  public isSignerSet = false
+
+  constructor(address: string, abi: any, private client: Harmony) {
+    super()
+    this.contract = this.client.contracts.createContract(abi, address)
+  }
+
+  async _getBalance(address: string, id?: string): Promise<BigNumber> {
+    if (!address) {
+      throw new Error('You have provide an address')
```
Comment (bzuker) [inline src/base-implementation.ts:35]: nit: `you have to`

File: src/base-implementation.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    this.code = code
+    this.data = data
+
+    Error.captureStackTrace(this, this.constructor)
+  }
+}
+
+export class BaseToken extends Base {
+  public contract: Contract
+  public isSignerSet = false
```
Comment (bzuker) [inline src/base-implementation.ts:26]: this should probably be private

File: src/hrc1155.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      const balance: BigNumber = await this._getBalance(address, id)
+      return balance.toNumber()
+    } catch (error) {
+      return logger.throwError('bad result from backend', Logger.errors.SERVER_ERROR, {
+        method: 'balanceOf',
+        params: address,
+        error,
+      })
+    }
+  }
```
Comment (bzuker) [inline src/hrc1155.ts:26]: let's make `_getBalance` return a number directly. This method should only do `return this._getBalance` and leave the error handling to the underlying method

File: src/hrc721.ts [hunk 6]
```diff
@@ -42,7 +20,7 @@ export class HRC721 extends BaseHRC721 {
     }
 
     try {
-      const balance: BigNumber = await this.contract.methods.balanceOf(address).call()
+      const balance: BigNumber = await this._getBalance(address)
```
Comment (bzuker) [inline src/hrc721.ts:23]: same as above

File: src/base-implementation.ts [hunk 7]
```diff
@@ -0,0 +1,103 @@
+import { Wallet } from '@harmony-js/account'
+import { Contract } from '@harmony-js/contract'
+import { Harmony } from '@harmony-js/core'
+import { BigNumber, logger, constants } from 'ethers'
```
Comment (jarcodallo) [inline src/base-implementation.ts:4]: If we are going to use @harmony-js sdk, we should just use the individual packages from etherts. like in this case `import { BigNumber } from '@ethersproject/bignumber` or check in the @harmony-js sdk if they have something for handle BigNumbers, logger and other things that we'll need

File: src/base-implementation.ts [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+import { Logger } from 'ethers/lib/utils'
```
Comment (jarcodallo) [inline src/base-implementation.ts:5]: Why is logger twice?

File: src/base-implementation.ts (hunk 1 above)
Comment (AnnRiera) [reply src/base-implementation.ts:1]: Copy!

File: src/base-implementation.ts (hunk 3 above)
Comment (AnnRiera) [reply src/base-implementation.ts:35]: Oh, you're right, I didn't notice that

File: src/hrc721.ts [hunk 9]
```diff
@@ -1,39 +1,17 @@
-import { Wallet } from '@harmony-js/account'
-import { Contract } from '@harmony-js/contract'
 import { Harmony } from '@harmony-js/core'
 import { Transaction } from '@harmony-js/transaction'
 import { Unit } from '@harmony-js/utils'
 import { hexToNumber } from '@harmony-js/utils'
 import { BigNumber, logger } from 'ethers'
```
Comment (jarcodallo) [inline src/hrc721.ts:5]: same here

File: src/base-implementation.ts (hunk 2 above)
Comment (AnnRiera) [reply src/base-implementation.ts:24]: Okay!

File: src/base-implementation.ts (hunk 1 above)
Comment (AnnRiera) [reply src/base-implementation.ts:1]: Copy!

File: src/base-implementation.ts (hunk 2 above)
Comment (AnnRiera) [reply src/base-implementation.ts:24]: Okay!

File: src/base-implementation.ts (hunk 3 above)
Comment (AnnRiera) [reply src/base-implementation.ts:35]: Oh, you're right, I didn't realize before, thanks!

File: src/hrc1155.ts (hunk 5 above)
Comment (AnnRiera) [reply src/hrc1155.ts:26]: Okay, I'll fix it!

File: src/base-implementation.ts (hunk 8 above)
Comment (AnnRiera) [reply src/base-implementation.ts:5]: Idk, I didn't add those dependencies, I'll change it right away!

File: src/base-token.ts [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+      return logger.throwError('bad result from backend', Logger.errors.SERVER_ERROR, {
+        params: { address, id },
+        error,
+      })
+    }
+  }
+
+  async setApprovalForAll(addressOperator: string, approved: boolean): Promise<any> {
+    return
+  }
```
Comment (jarcodallo) [inline src/base-token.ts:57]: Add a throw here to make sure this is implemented in the child. Something like "setApprovalForAll is not implemented"

**Review by jarcodallo — APPROVED** [review]
LGTM. Great work. Just a minor fix.

File: src/base-token.ts (hunk 10 above)
Comment (AnnRiera) [reply src/base-token.ts]: Sure!

**Review by jarcodallo — APPROVED** [review]

**Pull request merged**